clippy_lints/src/needless_cow.rs

```diff
@@ -0,0 +1,384 @@
+        }
+
+        let mut results = NeedlessCowAnalysis {
+            map: Map::from_filter(cx.tcx, mir, |ty| ty.peel_refs().is_str(), None),
+        }
+        .wrap()
+        .into_engine(cx.tcx, mir)
+        .iterate_to_fixpoint();
+
+        results.visit_reachable_with(
+            mir,
+            &mut NeedlessCowVisitor {
+                cx,
+                cow_locals_to_assignments: &mut cow_locals_to_assignments,
+            },
+        );
+
+        for assignments in cow_locals_to_assignments.values() {
+            if assignments.len() == 1 {
+                continue;
+            }
+            let Some(first) = assignments.first() else {
+                continue;
+            };
+
+            span_lint_hir_and_then(
+                cx,
+                NEEDLESS_COW,
+                mir.source_scopes[first.source_info().scope]
+                    .local_data
+                    .as_ref()
+                    .assert_crate_local()
+                    .lint_root,
+                assignments.iter().map(|a| a.source_info().span).collect_vec(),
+                "usage of `Cow` where the `Borrowed` variant is only ever used for empty strings",
+                |diag| {
+                    diag.help("remove the `Cow` and use `String::new` instead, as it allocates no memory");
+
+                    for into_or_from in assignments.iter().filter(|a| matches!(a, CowAssignment::IntoOrFrom(_))) {
+                        diag.span_note(into_or_from.source_info().span, "this implicitly constructs `Borrowed`");
+                    }
+                },
+            );
+        }
+    }
+}
+
+type CowAssignments = Vec<CowAssignment>;
+
+#[derive(Clone, Copy)]
+enum CowAssignment {
+    Borrowed(SourceInfo),
+    Owned(SourceInfo),
+    IntoOrFrom(SourceInfo),
+}
+
+impl CowAssignment {
+    fn source_info(self) -> SourceInfo {
+        match self {
+            Self::Borrowed(s) | Self::Owned(s) | Self::IntoOrFrom(s) => s,
+        }
+    }
+}
+
+struct NeedlessCowAnalysis {
+    map: Map,
+}
+
+impl<'tcx> ValueAnalysis<'tcx> for NeedlessCowAnalysis {
+    type Value = FlatSet<ConstantKind<'tcx>>;
+
+    const NAME: &'static str = "NeedlessCowAnalysis";
+
+    fn map(&self) -> &Map {
+        &self.map
+    }
+
+    fn handle_assign(&self, target: Place<'tcx>, rvalue: &Rvalue<'tcx>, state: &mut State<Self::Value>) {
+        state.flood(target.as_ref(), &self.map);
+
+        let Some(target_idx) = self.map.find(target.as_ref()) else {
+            return;
+        };
+
+        let result = match rvalue {
+            Rvalue::Use(Operand::Constant(box constant)) => ValueOrPlace::Value(FlatSet::Elem(constant.literal)),
+            Rvalue::Use(Operand::Move(place) | Operand::Copy(place))
+            | Rvalue::Ref(_, _, place)
+            | Rvalue::CopyForDeref(place) => {
+                if (place.projection.is_empty() || place.is_indirect_first_projection() && place.projection.len() == 1)
```

Comment:
We can change this to always clearing projections, as all projections are either nops or do nothing for `str`:

- `Deref`: `str` doesn't implement `Deref` and can't by external code;
- `Field`: `str` has no fields since it's a primitive;
- `Index`: Indexing an empty `str` will always panic;
- `ConstantIndex`: Same as above;
- `Subslice`: Same as above;
- `Downcast`: `str` is not an enum or generator;
- `OpaqueCast`: Since we only track `str` it must already be a concrete type;

However, I'm not sure if this is what we *should* do or if it's correct. Am I missing something that allows us to track the value for these places easier?